eventsourcing/src/test/java/org/axonframework/eventsourcing/eventstore/StorageEngineTestSuite.java

```diff
@@ -256,58 +229,200 @@ void transactionRejectedWhenConcurrentlyCreatedTransactionIsCommittedFirst() {
+                   .thenCompose(AppendTransaction::commit)
+                   .get(5, TimeUnit.SECONDS);
+
+        StreamingCondition testCondition =
+                StreamingCondition.conditionFor(new GlobalSequenceTrackingToken(10), TEST_CRITERIA);
+
+        MessageStream<EventMessage<?>> result = testSubject.stream(testCondition);
+
+        try {
+            assertTrue(result.next().isEmpty());
+        } finally {
+            result.close();
+        }
     }
 
     @Test
     void eventsPublishedAreIncludedInOpenStreams() throws Exception {
         testSubject.appendEvents(AppendCondition.none(),
                                  taggedEventMessage("event-0", TEST_CRITERIA_TAGS),
                                  taggedEventMessage("event-1", TEST_CRITERIA_TAGS))
-                   .thenCompose(AppendTransaction::commit).get(5, TimeUnit.SECONDS);
+                   .thenCompose(AppendTransaction::commit)
+                   .get(5, TimeUnit.SECONDS);
 
         MessageStream<EventMessage<?>> stream = testSubject.tailToken()
-                                                           .thenApply(StreamingCondition::startingFrom).thenApply(
-                        testSubject::stream).get(5, TimeUnit.SECONDS);
+                                                           .thenApply(StreamingCondition::startingFrom)
+                                                           .thenApply(testSubject::stream)
+                                                           .get(5, TimeUnit.SECONDS);
 
         AssertUtils.assertWithin(1, TimeUnit.SECONDS, () -> assertTrue(stream.next().isPresent()));
         AssertUtils.assertWithin(1, TimeUnit.SECONDS, () -> assertTrue(stream.next().isPresent()));
 
-        testSubject.appendEvents(AppendCondition.none(),
-                                 taggedEventMessage("event-3", TEST_CRITERIA_TAGS))
-                   .thenCompose(AppendTransaction::commit).get(5, TimeUnit.SECONDS);
-
-        AssertUtils.assertWithin(1,
-                                 TimeUnit.SECONDS,
-                                 () -> assertTrue(stream.hasNextAvailable()));
-
-        assertEquals("event-3",
-                     stream.next().map(e -> (String) e.message().getPayload())
-                           .orElse("none"));
+        testSubject.appendEvents(AppendCondition.none(), taggedEventMessage("event-3", TEST_CRITERIA_TAGS))
+                   .thenCompose(AppendTransaction::commit)
+                   .get(5, TimeUnit.SECONDS);
+
+        await(
+                "Await until stream contains newly appended events"
+        ).atMost(Duration.ofSeconds(1))
+         .pollDelay(Duration.ofMillis(100))
+         .untilAsserted(() -> assertTrue(stream.hasNextAvailable()));
+
+        assertEquals(
+                "event-3",
+                stream.next()
+                      .map(e -> {
+                          if (e.message().getPayload() instanceof String payload) {
```

Comment:
Why do you need to guess the payload type here? :thinking: Does this depend on event store serialization?